Resolve Issue 673: no longer checks subclass type annotations when exiting super's `__init__`

ckatsube opened the pull request:
<!-- Provide a summary of your changes in the Pull Request Title above. -->
<!-- If this is a work in progress (not yet ready to be merged), make this a draft pull request. -->

## Motivation and Context

<!-- Why is this pull request required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here: -->
<!-- https://docs.github.com/en/github/managing-your-work-on-github/managing-your-work-with-issues-and-pull-requests/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword -->

This PR aims to resolve #673, as super classes are not expected to have initialized the subclass' instance attributes nor are subclasses expected to instantiate them before calling `super().__init__`.

## Your Changes

<!-- Describe your changes here. -->
**Description**: the two commits are actually two different versions proposed to resolve this issue.
1st commit (3e3f7fc): after calling the instance's wrapped method, the type annotation check is run on the class in `instance`'s hierarchy that the wrapped method was defined in. I thought this was appropriate as a non-overriden super class method is not responsible for modifying any of its subclass' instance attributes. Even if said subclass overrode one of its super's instance attributes, by LSP the behavior when using the super's method should be the same as if it were just an instance of the super. This does somewhat lean on the wrapped class utilizing proper OOP design, which leads to the next version.
2nd commit (bfcad20): the post-method call type check is only run if the wrapped method is the public/outermost method (it's not an overriden method). I also thought this could be appropriate because an overriden method would be called from the method that overrides it, so ultimately checking all type annotations after the outermost method is finished would also be after all the super calls but before the function exists back to the client code.

**Type of change** (select all that apply):
<!-- Put an `x` in all the boxes that apply. -->
<!-- Remove any lines that do not apply. --> 

- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Testing

<!-- Please describe in detail how you tested this pull request. -->
<!-- This can include tests you added and manual testing. -->

Confirmed that the pytest suite ran.
Confirmed that the MR provided in the issue passes.

Couldn't quite wrap my head around how to build a test suite for these changes, but one may get added if I figure it out.

## Questions and Comments (if applicable)

<!-- Ask any questions you have for the maintainers of this project regarding this PR. -->
<!-- Please describe the steps you have already taken to find the answer to your question. -->
<!-- This will ensure that we can give you clear and relevant advice. -->
<!-- If you have additional comments add them here as well. -->

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have verified that the TravisCI tests have passed. <!-- (check after opening pull request) -->
- [x] I have reviewed the test coverage changes reported on Coveralls. <!-- (check after opening pull request) -->

ckatsube commented:
Changelog:
added type annotations to the `tests/test_subclass_contracts.py` classes. Confirmed that adding them would throw the error in #673 and that the provided change will fix it.

david-yz-liu commented:
@ckatsube great research on this, and I really liked the two different approaches you mentioned. They get at two different limitations of our current approach:

1. An instance method should only be checking type annotations for the class where the method was defined and superclasses of that class, but not subclasses.
2. When `__init__` is called from client code, we should be able to use helper methods (from superclasses or from the same class) to initialize attributes where the type annotations (and in fact all representation invariants) aren't checked yet. We should only check these when the original call to `__init__` returns, and not earlier.

I think I'll get you to address these in two separate PRs. For this current PR, you can tackle the first item by using a variation of the approach from the first commit. But to make it cleaner, modify the interface for `_instance_method_wrapper` so that it takes in the `klass` value from the place where it's called, rather than `rep_invariants`. This way you'll be able to pass `klass` directly to the `_check_class_type_annotations` helper to retrieve its type annotations.

Also, the change you made to the test cases is a good one, but remove the blank line between the class docstring and the attribute type annotations that you added.

david-yz-liu commented:
@ckatsube one thing I forgot to mention about testing: this change should affect all superclass methods, not just `__init__`. So you should be able to add tests that check that when a superclass method is called inside a subclass method, the superclass method doesn't check the attribute types or representation invariants of the subclass.

ckatsube commented:
The changes were a lot more than I expected so the changelog is a little more extensive than usual.

Changelog:

`tests/test_subclass_contracts.py`:
- removed space between docstring and type annotations

`python_ta/contracts/__init__.py`:
- updated `_instance_method_wrapper` to take in the class that defined the wrapped method by replacing `rep_invariants` (which can still be accessed from the class itself.
  - extracted the `rep_invariants is None` check to the call-point. There is a slight difference in that before, `check_contracts` decorator (which was returned by `_instance_method_wrapper`) was assigned directly as the method instead of wrapping the method--now `check_contracts(value)` (where `value` is the method to be wrapped) is assigned instead. I believe the previous behavior was unintentional as the decorator was being assigned directly without actually decorating anything.
  - changed `klass_mod` in this changed helper to take the global context of the defined class, not the instance's class. I believe this was the intended behavior (as this was how it was done in the `new_setattr`) but the earlier versions did not have access to the true class context and just used the lesser ideal instance's class.
- replaced `rep_invariants` parameter of `_check_invariants` for `klass`. Seemed redundant to force the client code in `_instance_method_wrapper` to re-access the invariants or to pass it through `_instance_method_wrapper` from its `add_class_invariants` parent caller when the invariants can just be accessed directly from the class. There is the argument that you might want to pass in invariants that are not directly tied to a singular class and maybe a mix of classes, but the current code base didn't seem to demand that kind of behavior.
  - this change did require `new_setattr` in `add_class_invariants` to pass in the class (which it already had access to) instead of the invariants.

ckatsube commented:
> @ckatsube one thing I forgot to mention about testing: this change should affect all superclass methods, not just `__init__`. So you should be able to add tests that check that when a superclass method is called inside a subclass method, the superclass method doesn't check the attribute types or representation invariants of the subclass.

Alright. I'll get working on this as well!

david-yz-liu commented on python_ta/contracts/__init__.py:
```diff
@@ -143,29 +144,28 @@ def _check_function_contracts(wrapped, instance, args, kwargs):
 
-def _instance_method_wrapper(wrapped, rep_invariants=None):
-    if rep_invariants is None:
-        return check_contracts
```
It's good that you flagged this, although the fix should be a bit different. It's okay if there are no representation invariants, we should still execute the `wrapper` so that the type annotations are checked. So you should revert the change on line 102/104, _keep_ the removal of these two lines, and see my comment below.

david-yz-liu commented on python_ta/contracts/__init__.py:
```diff
@@ -143,29 +144,28 @@ def _check_function_contracts(wrapped, instance, args, kwargs):
-def _instance_method_wrapper(wrapped, rep_invariants=None):
-    if rep_invariants is None:
-        return check_contracts
+def _instance_method_wrapper(wrapped, klass: type):
```
Add type annotations for `wrapped` and return type as well.

david-yz-liu commented on python_ta/contracts/__init__.py:
```diff
@@ -143,29 +144,28 @@ def _check_function_contracts(wrapped, instance, args, kwargs):
     """Check that the type annotations for the class still hold.
     """
-    klass = instance.__class__
+    assert isinstance(instance, klass)
```
Make this a precondition instead.

david-yz-liu commented on python_ta/contracts/__init__.py:
```diff
@@ -178,9 +178,14 @@ def _check_class_type_annotations(instance: Any) -> None:
+def _check_invariants(instance, klass: type, global_scope: dict) -> None:
     """Check that the representation invariants for the instance are satisfied.
+
+    Precondition:
```
Remove this precondition. Instead, use `getattr(klass, '__representation_invariants__', set())` to use a default value of an empty set.

david-yz-liu commented on tests/test_subclass_contracts.py:
```diff
@@ -29,19 +33,53 @@ class Developer(Employee):
+        self.wage_per_class = wage_per_class
+        self.currently_teaching = currently_teaching
+
+    def teach_new_classes(self, currently_teaching):
```
This is a good test, but you should also include another version with the opposite execution order: `self.change_wages` is called first, then `self.currently_teaching` is reassigned. This is what I meant by my earlier comment: `self.change_wages` will violate a representation invariant for `Teacher` but not `Employee`, and so no error should be raised in this case either.

david-yz-liu commented on tests/test_subclass_contracts.py:
```diff
@@ -128,3 +171,35 @@ def test_change_teamlead_team_invalid(teamlead):
     assert 'len(self.team) > 0' in msg
+
+
+def test_teacher_rep_violatable(pe_bio_teacher):
```
I think this test and `test_method_violates_class_invariant` are redundant with the ones using the `Developer` class.

ckatsube replied:
I'm making a case that I want to keep one of the two as a test to check that the class' invariant is written properly and actually works so it isn't being skipped and throwing false positives.

david-yz-liu replied:
I see what you're saying, but I do not think this is important enough to justify the additional code. However, instead of creating a new subclass you can reuse the existing subclass (`Developer`) and just create a new method and test with that.

That said, next time I ask you to do something, you should do it even if you want to argue against it. That way I can respond and merge in your pull request at once, rather than waiting another cycle.

ckatsube replied:
Understood. I'll make sure to do that in the future

david-yz-liu commented:
@ckatsube btw there's a small merge conflict because of #721, your change still applies, the function call just moved a bit.
